### ISSUE TITLE
Using a refresh token to create an azidentity credential

### ISSUE THREAD

**Issue by ogazitt** [Azure/azure-sdk-for-go#20299]
I have an Azure AD app that has a client ID and client secret. I'd like to have my users use the OAuth2 flow to authorize the application for their tenant.  I can successfully get a refresh token using the standard OAuth2 flow, and store that refresh token so that my app can talk to the graph API on a schedule (hourly).

I was assuming there would be an azidentity credential type that supports a bearer token (or a refresh token that can be traded for a short-lived access token).  The closest I can find is the ClientSecretCredential which takes a tenant, client ID, and client secret. But that defeats the OAuth2 flow of obtaining a refresh token once, and not having to obtain the client ID and client secret every time I need to get a short-lived access token.

Instead, ideally I'd like to store a refresh token for every tenant that has authorized my application, and initialize a credential with the refresh token, which can trade that for a short-lived access token. 

Is this possible to do with azidentity?

**Comment by chlowell** [comment]
Which authentication flow does your app use? Your description sounds like the on-behalf-of flow, which is supported in the latest azidentity beta by [`OnBehalfOfCredential`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azidentity@v1.3.0-beta.3#OnBehalfOfCredential). Like other credential types, it handles the details of authentication and token caching internally.

**Comment by ogazitt** [comment]
Thanks for the response Charles.

My SaaS app provides the ability for a user to connect various identity
providers (via PATs or an OAuth2 flow where available). In the case of AAD,
ideally I'd like to use the OAuth2 authorization code flow (on-behalf-of),
described here <https://learn.microsoft.com/en-us/graph/auth-v2-user>. I
want to split this up into two phases:

1. I have an existing node.js service that handles the OAuth2 initiation
and callback logic for the various providers I support. At the end of the
flow, I can store a short-lived access token and a long-lived refresh
token, as well as the user's tenant ID, into a tenant-specific secret
bucket.
2. Once an hour, I have a backend service (written in go) that needs to
query the graph API, and retrieve data on behalf of the user (in our case,
AAD users and groups). For other providers, this service gets the
tenant-specific secret (a PAT, or a refresh/access token), and queries the
identity provider for data using the access token.

My hope is that there is an azidentity credential type that can be
initialized with a tenant ID and refresh token, which can be used to obtain
an access token. I could then use this credential to create the
AzureIdentityAuthenticationProvider and create a GraphRequestAdapter using
that auth provider.

The OnBehalfOfCredential you linked to appears to require a clientID and
clientSecret, which is what I'm trying to avoid by storing a refresh token.

In this way, it doesn't appear all that different from the
ClientSecretCredential, which is what I'm using here
<https://github.com/aserto-dev/aserto-idp-plugin-azuread/blob/main/pkg/azureclient/azure.go>
(successfully). But the way my system is designed, I'm trying to store only
tenant-specific secrets for each one of my tenants (i.e. a refresh token /
tenant ID), which can be used on their own in order to authenticate to the
system I'm connecting to. If I also have to supply a ClientID and
ClientSecret (which are my app's secrets, not my customer's secrets), this
requires me to rework my provider model.

Thanks!

**Comment by chlowell** [comment]
Thanks for the details. This does look like a scenario for `OnBehalfOfCredential`, because you want users to delegate access to the backend application. The key difference between what you want to do and `ClientSecretCredential` is that `ClientSecretCredential` authenticates only the application itself and can't receive delegated access from users. (That's the nature of the underlying auth flow, not a restriction imposed by the Go API.)

azidentity handles token refreshing and refresh tokens internally to protect account security, among other reasons, so it doesn't have a credential type like the one you describe, and we don't plan to add one. If this is the only way forward for your application architecture, you could write a custom credential type to handle the refreshing step; the interface is [azcore.TokenCredential](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.3.1/internal/exported#TokenCredential).

**Comment by ogazitt** [comment]
Thanks, that's helpful, and that's likely what I'll end up doing. Store the refresh token, make a REST call to trade it for a short-lived access token, and wrap it in the `azcore.TokenCredential` interface, so that it exposes the token in `GetToken()`.

Is there code you can point me to in `azidentity` which I can borrow which calls the MS OAuth2 endpoints to trade a refresh token for an access token?

**Comment by chlowell** [comment]
For azidentity, refreshing is actually handled by [MSAL for Go](https://github.com/AzureAD/microsoft-authentication-library-for-go/blob/9a63cc219c3822f23f3a54c62c1261b729a50536/apps/internal/oauth/ops/accesstokens/accesstokens.go#L233), and the code there is more complex than you would need. The [Graph docs](https://learn.microsoft.com/graph/auth-v2-user#5-use-the-refresh-token-to-get-a-new-access-token) are easier to follow.

**Comment by ogazitt** [comment]
To close the loop, I implemented a `RefreshTokenCredential` which uses the refresh token flow to obtain a short-lived access token. Unfortunately the way that the graph APIs implement refresh grant requires both the client ID and client secret, so I still have to hold on to them.  I may as well just use the `ClientSecretCredential`.

**Comment by chlowell** [comment]
Great, I'm glad you're unblocked. I'll close this issue then. Please feel free to open another if something else comes up. As a side note, `ClientSecretCredential` uses the client credentials flow, which doesn't involve a refresh token because that flow authenticates the client application itself instead of impersonating a user.
